### PR TITLE
Host certificate lambda

### PR DESCRIPTION
**Pull request by vivianho**
Lambda is modified to allow the option to create host certificates.

### CODE REVIEW

**Comment by russell-lewis** [comment]
Thanks for the PR, and it is a great feature to add!  I've got a some feedback on it, that i'll write up next week.

**Comment by vivianho** [comment]
Any updates on this @russell-lewis ?

**Comment by russell-lewis** [comment]
Sorry about the delays, I had some other urgent issues to attend to last week and this week.  I'll try to get back to you next week.

File: bless/request/bless_request.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 
 # man 8 useradd
 USERNAME_PATTERN = re.compile('[a-z_][a-z0-9_-]*[$]?\Z')
+HOSTNAME_PATTERN = re.compile('[a-z0-9_.-]+')
```
Comment (russell-lewis) [inline bless/request/bless_request.py]: Do you have hostnames with an _ in them?
https://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_host_names

File: bless/aws_lambda/bless_lambda.py [hunk 2]
```diff
@@ -94,24 +108,41 @@ def lambda_handler(event, context=None, ca_private_key_password=None,
...
     cert = cert_builder.get_cert_file()
 
+    if certificate_type == SSHCertificateType.HOST:
+        remote_name = ', '.join(request.remote_hostnames)
+        bastion_ip = None
+    else:
+        remote_name = request.remote_username
+        bastion_ip = request.bastion_ip
+
     logger.info(
```
Comment (russell-lewis) [inline bless/aws_lambda/bless_lambda.py]: Wouldn't you rather have a different audit log message for host vs user certs?

File: bless/aws_lambda/bless_lambda.py [hunk 3]
```diff
@@ -11,14 +11,23 @@
...
-from bless.request.bless_request import BlessSchema
+    BLESS_CA_SECTION, CA_PRIVATE_KEY_FILE_OPTION, LOGGING_LEVEL_OPTION, CERTIFICATE_TYPE_OPTION
+from bless.request.bless_request import BlessUserSchema, BlessHostSchema
 from bless.ssh.certificate_authorities.ssh_certificate_authority_factory import \
     get_ssh_certificate_authority
 from bless.ssh.certificates.ssh_certificate_builder import SSHCertificateType
 from bless.ssh.certificates.ssh_certificate_builder_factory import get_ssh_certificate_builder
 
 
+def get_certificate_type(certificate_type_option):
```
Comment (russell-lewis) [inline bless/aws_lambda/bless_lambda.py]: This validation is good, but I think you should move it to BlessConfig's constructor.

File: bless/aws_lambda/bless_lambda.py [hunk 4]
```diff
@@ -31,6 +40,7 @@ def lambda_handler(event, context=None, ca_private_key_password=None,
     decrypt.
     :param entropy_check: For local testing, if set to false, it will skip checking entropy and
     won't try to fetch additional random from KMS
+    :param certificate_type: Type of certificate to be generated
```
Comment (russell-lewis) [inline bless/aws_lambda/bless_lambda.py]: You document a parameter, but didn't create/use it.

**Comment by vivianho** [comment]
Thanks for your comments, @russell-lewis. I've been working on a lyft-specific host lambda (which has since diverged from this branch), but I can address the comments in this branch soon.

File: bless/config/bless_deploy_example.cfg [hunk 5]
```diff
@@ -1,13 +1,15 @@
 # This section and its options are optional
 [Bless Options]
 # Number of seconds +/- the issued time for the certificate to be valid
-certificate_validity_window_seconds = 120
+certificate_validity_seconds = 120
```
Comment (russell-lewis) [inline bless/config/bless_deploy_example.cfg]: I thought "validity window" was a useful config for short lived client certs.  However for host certs, this seems like it wouldn't really be ideal for your use case, assuming you issue host certs for far off valid_before range.

Would separate valid_before and valid_after offsets better suit your needs?  I'm guessing you don't have a need to issue valid after t=0 certs.

File: bless/request/bless_request.py [hunk 6]
```diff
@@ -51,14 +88,27 @@ def __init__(self, bastion_ip, bastion_user, bastion_user_ip, command, public_ke
...
         self.public_key_to_sign = public_key_to_sign
         self.remote_username = remote_username
 
-    def __eq__(self, other):
-        return self.__dict__ == other.__dict__
+
+class BlessHostRequest(BlessRequest):
+    def __init__(self, public_key_to_sign, remote_hostnames):
+        """
+        A BlessRequest must have the following key value pairs to be valid.
```
Comment (russell-lewis) [inline bless/request/bless_request.py]: A BlessHostRequest ..

File: bless/request/bless_request.py [hunk 7]
```diff
@@ -25,20 +26,56 @@ def validate_user(user):
...
+        enforced in the issued certificate.
+        """
+
+        self.public_key_to_sign = public_key_to_sign
+
+    def __eq__(self, other):
+        return self.__dict__ == other.__dict__
+
+
+class BlessUserRequest(BlessRequest):
```
Comment (russell-lewis) [inline bless/request/bless_request.py]: Update docstring on line 82.
A BlessUserRequest ...

File: bless/request/bless_request.py [hunk 8]
```diff
@@ -51,14 +88,27 @@ def __init__(self, bastion_ip, bastion_user, bastion_user_ip, command, public_ke
         :param public_key_to_sign: The id_rsa.pub that will be used in the SSH request.  This is
         enforced in the issued certificate.
         :param remote_username: The username on the remote server that will be used in the SSH
-        request.  This is enforced in the issued certificate.
```
Comment (russell-lewis) [inline bless/request/bless_request.py]: Why was this line removed?

**Comment by russell-lewis** [comment]
@vivianho Sounds good to me.  Let me know if you'd like me to take a look at that other branch so we can get/keep things aligned.

File: bless/config/bless_deploy_example.cfg (hunk 5 above)
Comment (vivianho) [inline bless/config/bless_deploy_example.cfg]: We have certificate validity set for 7 days but valid_before and valid_after could work as well. I could add them in a separate PR?
